server/server.go

```diff
@@ -851,7 +866,123 @@ func (s *Server) handleExit(ctx context.Context, event fsnotify.Event) {
 	if err := s.ContainerStateToDisk(ctx, c); err != nil {
 		log.Warnf(ctx, "Unable to write %s %s state to disk: %v", resource, c.ID(), err)
 	}
+	s.generateCRIEvent(ctx, c, types.ContainerEventType_CONTAINER_STOPPED_EVENT)
 	if err := os.Remove(event.Name); err != nil {
 		log.Warnf(ctx, "Failed to remove exit file: %v", err)
 	}
 }
+
+func (s *Server) getSandboxStatuses(ctx context.Context, sandboxID string) (*types.PodSandboxStatus, error) {
+	sandboxStatusRequest := &types.PodSandboxStatusRequest{PodSandboxId: sandboxID}
+	sandboxStatus, err := s.PodSandboxStatus(ctx, sandboxStatusRequest)
+
+	if isNotFound(err) {
+		return nil, err
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("getSandboxStatuses: %w", err)
+	}
+
+	return sandboxStatus.GetStatus(), nil
+}
+
+func (s *Server) getContainerStatuses(ctx context.Context, sandboxUid string) ([]*types.ContainerStatus, error) {
+	listContainerRequest := &types.ListContainersRequest{Filter: &types.ContainerFilter{LabelSelector: map[string]string{"io.kubernetes.pod.uid": sandboxUid}}}
+	containers, err := s.ListContainers(ctx, listContainerRequest)
+	if err != nil {
+		return []*types.ContainerStatus{}, err
+	}
+
+	containerStatuses := []*types.ContainerStatus{}
+	for _, cc := range containers.GetContainers() {
+		containerStatusRequest := &types.ContainerStatusRequest{ContainerId: cc.Id}
+		resp, err := s.ContainerStatus(ctx, containerStatusRequest)
+		if isNotFound(err) {
+			continue
+		}
+		if err != nil {
+			return []*types.ContainerStatus{}, err
+		}
+		containerStatuses = append(containerStatuses, resp.GetStatus())
+	}
+
+	return containerStatuses, nil
+}
+
+func (s *Server) getContainerStatusesFromSandboxId(ctx context.Context, sandboxId string) ([]*types.ContainerStatus, error) {
+	listContainerRequest := &types.ListContainersRequest{Filter: &types.ContainerFilter{PodSandboxId: sandboxId}}
+	containers, err := s.ListContainers(ctx, listContainerRequest)
+	if err != nil {
+		return []*types.ContainerStatus{}, err
+	}
+
+	containerStatuses := []*types.ContainerStatus{}
+	for _, cc := range containers.GetContainers() {
+		containerStatusRequest := &types.ContainerStatusRequest{ContainerId: cc.Id, Verbose: false}
+		resp, err := s.ContainerStatus(ctx, containerStatusRequest)
+		if isNotFound(err) {
+			continue
+		}
+		if err != nil {
+			return []*types.ContainerStatus{}, err
+		}
+		containerStatuses = append(containerStatuses, resp.GetStatus())
+	}
+
+	return containerStatuses, nil
+}
+
+func (s *Server) generateCRIEvent(ctx context.Context, container *oci.Container, eventType types.ContainerEventType) {
+	// returning no error if the Evented PLEG feature is not enabled
+	if !s.config.EnablePodEvents {
+		return
+	}
+	if err := s.Runtime().UpdateContainerStatus(ctx, container); err != nil {
```

Comment:
I am still pretty anxious about adding this call. this means for every single pod event there's a corresponding call to `runc state` and all the overhead of managing a new process. We really shouldn't need this call, and if we do, we're not reacting to a container exit correctly, or are sending the event too early